planet/clients/orders.py

```diff
@@ -271,18 +273,41 @@ async def download_asset(self,
                              progress_bar=progress_bar)
         return dl_path
 
+    @staticmethod # should this be an async function?
+    def calculate_checksum(origin_hash: str,
+                        filename: str,
+                        checksum: str):
+        """Calculate checksum and validate that it passes."""
+        if checksum == 'MD5':
+            hash_type = hashlib.md5
+        elif checksum == 'SHA256':
+            hash_type = hashlib.sha256
+        # Calculate returned hash
+        with open(filename, 'rb') as file_to_check:
+            json_data = file_to_check.read()
+            returned_hash = hash_type(json_data).hexdigest()
+            # Compare original hashkey in dict with calculated
+            if origin_hash != returned_hash:
```

Comment:
Here you're comparing the `origin_hash`, as it's written in the JSON file, to the `returned_hash`, which has been put through `hexdigest()`, so it's no comparing apples to apples. 

For instance, if you put `origin_hash = 'anOriginMd5'` through `hash_type(json_data).hexdigest()` you'll get `'39f5b5db96c26ca61e13b280c8b98b6e'`, so you're comparing `'anOriginMd5 == '39f5b5db96c26ca61e13b280c8b98b6e'`, which will return false.

So, you'll need to compare `hash_type(origin_json_data).hexdigest() == hash_type(otherfile_json_data).hexdigest()`

Reply:
Here's an idea:

In `download_order()` you can replace a bunch of your code with just a few lines, as now you're going to pass along a lot of the heavy lifting to `calculate_checksum()`. Try:

```
if checksum == 'md5':
    # Checksum Implementation
    # Get manifest filepath
    manifest_json = ' '.join([x for x in filenames if x.endswith('manifest.json')])
    with open(manifest_json, 'rb') as manifest:
        manifest_data = manifest.read()

    # For each file (not including manifest json file), calculate hash on its contents
    filenames_loop = [x for x in filenames if not x.endswith('manifest.json')]
    # filenames_loop = ' '.join([x for x in filenames if not x.endswith('manifest.json')])
    for filename in filenames_loop:
        self.calculate_checksum(origin_json_data= manifest_data,
                            filename= filename,
                            checksum= checksum)
```

And in `calculate_checksum()` try this:
```
def calculate_checksum(origin_json_data: str,
                    filename: str,
                    checksum: str):
    """Calculate checksum and validate that it passes."""
    if checksum == 'md5':
        hash_type = hashlib.md5
    elif checksum == 'sha256':
        hash_type = hashlib.sha256
    # Calculate returned hash
    with open(filename, 'rb') as file_to_check:
        json_data = file_to_check.read()
        origin_hash = hash_type(origin_json_data).hexdigest()
        returned_hash = hash_type(json_data).hexdigest()
        # Compare original hashkey in dict with calculated
        if origin_hash != returned_hash:
            print('origin_hash: ', origin_hash)
            print('returned_hash: ', returned_hash)
            raise exceptions.ClientError(
                'Checksum failed. File not correctly downloaded.')
        return f'({checksum}) checksum succesful: ({filename})'
```

Reply:
Now you're comparing apples to apples! 🍎 🍎

Reply:
hmmmm i'm not sure the hexdigest is the issue? I think it's different with strings, but that's needed to read the encoded data in the file so that I can see it as a string and compare it that way. Also, i'm trying to keep the checksum function as simple as possible (not incorporating the part of the function that loads the manifest) to keep things as simple as possible for testing as i was having issues with that previously.